File: apprain_barcode_scanner_ios/ios/Classes/ScannerPlatformView.swift
import AVFoundation
import Flutter
import UIKit
import Vision

/// Native iOS view that hosts AVFoundation camera preview + Vision barcode detection.
///
/// Key design decisions:
/// - Uses AVCaptureSession for direct camera control.
/// - Vision framework VNDetectBarcodesRequest for barcode detection (iOS 11+).
/// - Frame processing on a dedicated serial queue to avoid blocking main thread.
/// - Supports ROI-based scanning via Vision's regionOfInterest.
/// - Frame throttling via configurable maxFrameRate.
class ScannerPlatformView: NSObject, FlutterPlatformView,
    AVCaptureVideoDataOutputSampleBufferDelegate {

    // MARK: - Properties

    private let previewContainer = UIView()
    private var previewLayer: AVCaptureVideoPreviewLayer?

    // Camera
    private let captureSession = AVCaptureSession()
    private var currentDevice: AVCaptureDevice?
    private var videoOutput: AVCaptureVideoDataOutput?
    private let processingQueue = DispatchQueue(label: "com.apprain.barcode_scanner.processing", qos: .userInteractive)

    // Detection
    private var barcodeRequest: VNDetectBarcodesRequest?
    private var isProcessing = false
    private var isConfiguring = false
    private var lastAnalysisTimestamp: TimeInterval = 0

    // Image preprocessing (GPU-accelerated)
    private let ciContext = CIContext(options: [.useSoftwareRenderer: false])

    // Config
    private var useFrontCamera = false
    private var enableScanWindow = true
    private var scanWindowWidthRatio: Double = 0.75
    private var scanWindowHeightRatio: Double = 0.35
    private var scanWindowVerticalOffsetRatio: Double = 0.10
    private var enablePreprocessing = true
    private var maxFrameRate = 15
    private var enableTorch = false
    private var allowedSymbologies: [VNBarcodeSymbology] = []
    private var targetResolution: String = "hd720"

    // Event sinks
    private var scanEventSink: FlutterEventSink?
    private var cameraEventSink: FlutterEventSink?

    // Stream handlers (kept alive to maintain event sink references)
    private let scanStreamHandler = _EventStreamHandler()
    private let cameraStreamHandler = _EventStreamHandler()

    // MARK: - Init

    init(
        frame: CGRect,
        viewId: Int64,
        creationParams: [String: Any]?,
        scanEventChannel: FlutterEventChannel,
        cameraEventChannel: FlutterEventChannel
    ) {
        super.init()

        previewContainer.frame = frame
        previewContainer.backgroundColor = .black

        // Scan event channel — barcode detections
        scanStreamHandler.onSinkReady = { [weak self] sink in
            self?.scanEventSink = sink
        }
        scanStreamHandler.onSinkCancelled = { [weak self] in
            self?.scanEventSink = nil
        }
        scanEventChannel.setStreamHandler(scanStreamHandler)

        // Camera event channel — lifecycle events
        cameraStreamHandler.onSinkReady = { [weak self] sink in
            self?.cameraEventSink = sink
        }
        cameraStreamHandler.onSinkCancelled = { [weak self] in
            self?.cameraEventSink = nil
        }
        cameraEventChannel.setStreamHandler(cameraStreamHandler)
    }

    func view() -> UIView {
        return previewContainer
    }

    // MARK: - Public API

    func initialize(config: [String: Any]?) {
        applyConfig(config: config)
        initBarcodeDetector(config: config)
    }

    func startCamera() {
        processingQueue.async { [weak self] in
            guard let self = self else { return }
            self.setupCamera()

            if !self.captureSession.isRunning {
                self.captureSession.startRunning()
            }

            DispatchQueue.main.async {
                self.setupPreviewLayer()
                self.sendCameraEvent(type: "ready")
            }
        }
    }

    func stopCamera() {
        processingQueue.async { [weak self] in
            guard let self = self else { return }
            // Never call stopRunning while inside a configuration block
            guard !self.isConfiguring else { return }
            if self.captureSession.isRunning {
                self.captureSession.stopRunning()
            }
        }
    }

    func disposeResources() {
        // Perform all cleanup on the processing queue to avoid
        // calling stopRunning between beginConfiguration/commitConfiguration
        processingQueue.async { [weak self] in
            guard let self = self else { return }
            if self.captureSession.isRunning {
                self.captureSession.stopRunning()
            }
            self.captureSession.inputs.forEach { self.captureSession.removeInput($0) }
            self.captureSession.outputs.forEach { self.captureSession.removeOutput($0) }

            DispatchQueue.main.async {
                self.previewLayer?.removeFromSuperlayer()
                self.previewLayer = nil
                self.scanEventSink = nil
                self.cameraEventSink = nil
            }
        }
    }

    func setTorch(enabled: Bool) {
        guard let device = currentDevice, device.hasTorch else { return }
        try? device.lockForConfiguration()
        device.torchMode = enabled ? .on : .off
        device.unlockForConfiguration()
        enableTorch = enabled
    }

    func setZoom(_ zoomLevel: CGFloat) {
        guard let device = currentDevice else { return }
        let maxZoom = min(device.activeFormat.videoMaxZoomFactor, 10.0)
        let zoom = 1.0 + zoomLevel * (maxZoom - 1.0)
        try? device.lockForConfiguration()
        device.videoZoomFactor = max(1.0, min(zoom, maxZoom))
        device.unlockForConfiguration()
    }

    func setFocusPoint(x: CGFloat, y: CGFloat) {
        guard let device = currentDevice else { return }
        guard device.isFocusPointOfInterestSupported else { return }

        let point = CGPoint(x: x, y: y)
        try? device.lockForConfiguration()
        device.focusPointOfInterest = point
        device.focusMode = .autoFocus
        if device.isExposurePointOfInterestSupported {
            device.exposurePointOfInterest = point
            device.exposureMode = .autoExpose
        }
        device.unlockForConfiguration()
    }

    func switchCamera() {
        useFrontCamera.toggle()
        if captureSession.isRunning {
            stopCamera()
            // Small delay to allow session cleanup
            DispatchQueue.main.asyncAfter(deadline: .now() + 0.2) { [weak self] in
                self?.startCamera()
            }
        }
    }

    func updateConfig(config: [String: Any]?) {
        applyConfig(config: config)
    }

    // MARK: - Private Setup

    private func applyConfig(config: [String: Any]?) {
        guard let config = config else { return }
        useFrontCamera = config["useFrontCamera"] as? Bool ?? useFrontCamera
        enableScanWindow = config["enableScanWindow"] as? Bool ?? enableScanWindow
        scanWindowWidthRatio = config["scanWindowWidthRatio"] as? Double ?? scanWindowWidthRatio
        scanWindowHeightRatio = config["scanWindowHeightRatio"] as? Double ?? scanWindowHeightRatio
        scanWindowVerticalOffsetRatio = config["scanWindowVerticalOffsetRatio"] as? Double ?? scanWindowVerticalOffsetRatio
        enablePreprocessing = config["enablePreprocessing"] as? Bool ?? enablePreprocessing
        maxFrameRate = config["maxFrameRate"] as? Int ?? maxFrameRate
        enableTorch = config["enableTorch"] as? Bool ?? enableTorch
        targetResolution = config["targetResolution"] as? String ?? targetResolution
    }

    private func initBarcodeDetector(config: [String: Any]?) {
        let formats = (config?["allowedFormats"] as? [String]) ?? []
        allowedSymbologies = formats.compactMap { formatStringToSymbology($0) }

        barcodeRequest = VNDetectBarcodesRequest { [weak self] request, error in
            guard let self = self else { return }
            if let error = error {
                print("⚠️ Vision barcode error: \(error.localizedDescription)")
                return
            }

            guard let results = request.results as? [VNBarcodeObservation],
                  !results.isEmpty else { return }

            let scanResults: [[String: Any]] = results.compactMap { observation in
                guard let payload = observation.payloadStringValue else { return nil }

                // ROI filtering — check if barcode center is within scan window
                if self.enableScanWindow {
                    let centerX = observation.boundingBox.midX
                    let centerY = observation.boundingBox.midY

                    let windowLeft = (1.0 - self.scanWindowWidthRatio) / 2.0
                    let windowRight = 1.0 - windowLeft
                    let windowCenterY = 0.5 + self.scanWindowVerticalOffsetRatio
                    let windowTop = windowCenterY - self.scanWindowHeightRatio / 2.0
                    let windowBottom = windowCenterY + self.scanWindowHeightRatio / 2.0

                    guard centerX >= windowLeft && centerX <= windowRight &&
                          centerY >= windowTop && centerY <= windowBottom else {
                        return nil
                    }
                }

                // Convert corners from Vision coords (bottom-left origin)
                // to Flutter coords (top-left origin): flip Y axis
                let corners: [Double] = observation.topLeft.flippedXY()
                    + observation.topRight.flippedXY()
                    + observation.bottomRight.flippedXY()
                    + observation.bottomLeft.flippedXY()

                // Bounding box in normalized coords (0-1) for Flutter overlay
                let bb = observation.boundingBox
                let boundingBox: [String: Double] = [
                    "left": Double(bb.origin.x),
                    "top": Double(1.0 - bb.origin.y - bb.size.height),
                    "width": Double(bb.size.width),
                    "height": Double(bb.size.height),
                ]

                return [
                    "rawValue": payload,
                    "format": self.symbologyToString(observation.symbology),
                    "corners": corners,
                    "confidence": Double(observation.confidence),
                    "boundingBox": boundingBox,
                ]
            }

            if !scanResults.isEmpty {
                DispatchQueue.main.async {
                    self.scanEventSink?(scanResults)
                }
            }
        }

        if !allowedSymbologies.isEmpty {
            barcodeRequest?.symbologies = allowedSymbologies
        }
    }

    private func setupCamera() {
        isConfiguring = true
        captureSession.beginConfiguration()

        // Remove existing inputs/outputs
        captureSession.inputs.forEach { captureSession.removeInput($0) }
        captureSession.outputs.forEach { captureSession.removeOutput($0) }

        captureSession.sessionPreset = sessionPresetForResolution()

        // Select camera
        let position: AVCaptureDevice.Position = useFrontCamera ? .front : .back
        guard let device = AVCaptureDevice.default(
            .builtInWideAngleCamera, for: .video, position: position
        ) else {
            sendCameraEvent(type: "error", message: "Camera device not available")
            return
        }
        currentDevice = device

        // Configure device for continuous autofocus
        try? device.lockForConfiguration()
        if device.isFocusModeSupported(.continuousAutoFocus) {
            device.focusMode = .continuousAutoFocus
        }
        if device.isExposureModeSupported(.continuousAutoExposure) {
            device.exposureMode = .continuousAutoExposure
        }
        device.unlockForConfiguration()

        // Add input
        guard let input = try? AVCaptureDeviceInput(device: device) else {
            sendCameraEvent(type: "error", message: "Could not create camera input")
            return
        }

        if captureSession.canAddInput(input) {
            captureSession.addInput(input)
        }

        // Add video output
        let output = AVCaptureVideoDataOutput()
        output.alwaysDiscardsLateVideoFrames = true
        output.videoSettings = [
            kCVPixelBufferPixelFormatTypeKey as String: kCVPixelFormatType_32BGRA,
        ]
        output.setSampleBufferDelegate(self, queue: processingQueue)
        videoOutput = output

        if captureSession.canAddOutput(output) {
            captureSession.addOutput(output)
        }

        // Set video orientation on the output connection
        if let connection = output.connection(with: .video) {
            if connection.isVideoOrientationSupported {
                connection.videoOrientation = currentVideoOrientation()
            }
        }

        // Apply torch if needed
        if enableTorch {
            setTorch(enabled: true)
        }

        captureSession.commitConfiguration()
        isConfiguring = false
    }

    private func setupPreviewLayer() {
        previewLayer?.removeFromSuperlayer()

        let layer = AVCaptureVideoPreviewLayer(session: captureSession)
        layer.videoGravity = .resizeAspectFill
        layer.frame = previewContainer.bounds

        // Set preview orientation to match device
        if let connection = layer.connection, connection.isVideoOrientationSupported {
            connection.videoOrientation = currentVideoOrientation()
        }

        previewContainer.layer.addSublayer(layer)
        previewLayer = layer
    }

    /// Maps the current interface orientation to AVCaptureVideoOrientation.
    private func currentVideoOrientation() -> AVCaptureVideoOrientation {
        let scenes = UIApplication.shared.connectedScenes
        let windowScene = scenes.first(where: { $0 is UIWindowScene }) as? UIWindowScene
        let orientation = windowScene?.interfaceOrientation ?? .landscapeRight

        switch orientation {
        case .portrait:
            return .portrait
        case .portraitUpsideDown:
            return .portraitUpsideDown
        case .landscapeLeft:
            return .landscapeLeft
        case .landscapeRight:
            return .landscapeRight
        @unknown default:
            return .landscapeRight
        }
    }

    /// Maps device interface orientation to CGImagePropertyOrientation.
    ///
    /// Vision framework needs this to correctly interpret pixel buffer
    /// coordinates relative to the device's current screen orientation.
    /// Without this, coordinates are returned in the sensor's native
    /// portrait orientation — causing X/Y mismatch on iPad landscape.
    private func currentCGImageOrientation() -> CGImagePropertyOrientation {
        let scenes = UIApplication.shared.connectedScenes
        let windowScene = scenes.first(where: { $0 is UIWindowScene }) as? UIWindowScene
        let orientation = windowScene?.interfaceOrientation ?? .landscapeRight

        switch orientation {
        case .portrait:
            return .right
        case .portraitUpsideDown:
            return .left
        case .landscapeLeft:
            return .down
        case .landscapeRight:
            return .up
        @unknown default:
            return .up
        }
    }

    /// Maps the targetResolution config string to an AVCaptureSession.Preset.
    private func sessionPresetForResolution() -> AVCaptureSession.Preset {
        switch targetResolution {
        case "fullHd1080":
            return .hd1920x1080
        case "uhd4k":
            return .hd4K3840x2160
        default: // "hd720"
            return .hd1280x720
        }
    }

    // MARK: - AVCaptureVideoDataOutputSampleBufferDelegate

    func captureOutput(
        _ output: AVCaptureOutput,
        didOutput sampleBuffer: CMSampleBuffer,
        from connection: AVCaptureConnection
    ) {
        // Frame throttling
        let now = CACurrentMediaTime()
        let minInterval = maxFrameRate > 0 ? 1.0 / Double(maxFrameRate) : 0
        guard now - lastAnalysisTimestamp >= minInterval else { return }

        // Single-frame gate
        guard !isProcessing else { return }
        isProcessing = true
        lastAnalysisTimestamp = now

        guard let pixelBuffer = CMSampleBufferGetImageBuffer(sampleBuffer) else {
            isProcessing = false
            return
        }

        guard let request = barcodeRequest else {
            isProcessing = false
            return
        }

        // ── Pass 1: Raw frame ──────────────────────────────────────
        let orientation = currentCGImageOrientation()
        let rawHandler = VNImageRequestHandler(
            cvPixelBuffer: pixelBuffer,
            orientation: orientation,
            options: [:]
        )
        do {
            try rawHandler.perform([request])
            if let results = request.results, !results.isEmpty {
                // Raw frame succeeded — skip preprocessing
                isProcessing = false
                return
            }
        } catch {
            // Fall through to preprocessed pass
        }

        // ── Pass 2: Preprocessed frame ─────────────────────────────
        // Apply contrast enhancement + sharpening for small barcodes
        guard enablePreprocessing else {
            isProcessing = false
            return
        }

        if let enhancedBuffer = preprocessFrame(pixelBuffer) {
            let enhancedHandler = VNImageRequestHandler(
                cvPixelBuffer: enhancedBuffer,
                orientation: orientation,
                options: [:]
            )
            do {
                try enhancedHandler.perform([request])
            } catch {
                print("⚠️ Vision preprocessed request failed: \(error.localizedDescription)")
            }
        }

        isProcessing = false
    }

    // MARK: - Image Preprocessing

    /// Enhances the pixel buffer for better small barcode detection.
    ///
    /// Pipeline: Exposure adjust → Contrast stretch → Unsharp mask.
    /// Uses GPU-accelerated CIFilter chain for real-time performance.
    private func preprocessFrame(_ pixelBuffer: CVPixelBuffer) -> CVPixelBuffer? {
        let ciImage = CIImage(cvPixelBuffer: pixelBuffer)

        // 1) Exposure adjustment — brighten underexposed areas
        guard let exposureFilter = CIFilter(
            name: "CIExposureAdjust",
            parameters: [
                kCIInputImageKey: ciImage,
                "inputEV": 0.3,
            ]
        ), let exposedImage = exposureFilter.outputImage else {
            return nil
        }

        // 2) Contrast enhancement — make barcode edges stand out
        guard let contrastFilter = CIFilter(
            name: "CIColorControls",
            parameters: [
                kCIInputImageKey: exposedImage,
                "inputContrast": 1.4,
                "inputSaturation": 0.0,  // Grayscale for barcode detection
            ]
        ), let contrastedImage = contrastFilter.outputImage else {
            return nil
        }

        // 3) Sharpening — enhance fine details in small codes
        guard let sharpenFilter = CIFilter(
            name: "CIUnsharpMask",
            parameters: [
                kCIInputImageKey: contrastedImage,
                "inputRadius": 2.5,
                "inputIntensity": 0.8,
            ]
        ), let sharpenedImage = sharpenFilter.outputImage else {
            return nil
        }

        // Render back to a new pixel buffer
        var outputBuffer: CVPixelBuffer?
        let width = CVPixelBufferGetWidth(pixelBuffer)
        let height = CVPixelBufferGetHeight(pixelBuffer)
        let attrs: [String: Any] = [
            kCVPixelBufferCGImageCompatibilityKey as String: true,
            kCVPixelBufferCGBitmapContextCompatibilityKey as String: true,
        ]
        CVPixelBufferCreate(
            kCFAllocatorDefault,
            width, height,
            kCVPixelFormatType_32BGRA,
            attrs as CFDictionary,
            &outputBuffer
        )

        guard let output = outputBuffer else { return nil }
        ciContext.render(sharpenedImage, to: output)
        return output
    }

    // MARK: - Helpers

    private func sendCameraEvent(type: String, message: String? = nil) {
        var event: [String: Any] = ["type": type]
        if let message = message {
            event["message"] = message
        }
        DispatchQueue.main.async { [weak self] in
            self?.cameraEventSink?(event)
        }
    }

    // MARK: - Format Mapping

    private func formatStringToSymbology(_ format: String) -> VNBarcodeSymbology? {
        switch format.uppercased() {
        case "QR_CODE": return .qr
        case "CODE_128": return .code128
        case "CODE_39": return .code39
        case "CODE_93": return .code93
        case "EAN_13": return .ean13
        case "EAN_8": return .ean8
        case "UPC_E": return .upce
        case "DATA_MATRIX": return .dataMatrix
        case "PDF417": return .pdf417
        case "AZTEC": return .aztec
        case "CODABAR": return .codabar
        case "ITF": return .itf14
        default: return nil
        }
    }

    private func symbologyToString(_ symbology: VNBarcodeSymbology) -> String {
        switch symbology {
        case .qr: return "QR_CODE"
        case .code128: return "CODE_128"
        case .code39: return "CODE_39"
        case .code93: return "CODE_93"
        case .ean13: return "EAN_13"
        case .ean8: return "EAN_8"
        case .upce: return "UPC_E"
        case .dataMatrix: return "DATA_MATRIX"
        case .pdf417: return "PDF417"
        case .aztec: return "AZTEC"
        case .codabar: return "CODABAR"
        case .itf14: return "ITF"
        default: return "UNKNOWN"
        }
    }
}

// MARK: - CGPoint Extension

private extension CGPoint {
    func asList() -> [Double] {
        return [Double(x), Double(y)]
    }

    /// Returns coordinates with Y flipped (1 - y) for Vision → Flutter conversion.
    func flippedY() -> [Double] {
        return [Double(x), Double(1.0 - y)]
    }

    /// Returns coordinates with both X and Y flipped for Vision → Flutter conversion.
    func flippedXY() -> [Double] {
        return [Double(1.0 - x), Double(1.0 - y)]
    }
}

// MARK: - Event Stream Handler

/// Lightweight FlutterStreamHandler that forwards the event sink via closures.
private class _EventStreamHandler: NSObject, FlutterStreamHandler {
    var onSinkReady: ((@escaping FlutterEventSink) -> Void)?
    var onSinkCancelled: (() -> Void)?

    func onListen(
        withArguments arguments: Any?,
        eventSink events: @escaping FlutterEventSink
    ) -> FlutterError? {
        onSinkReady?(events)
        return nil
    }

    func onCancel(withArguments arguments: Any?) -> FlutterError? {
        onSinkCancelled?()
        return nil
    }
}

File: apprain_barcode_scanner_android/android/src/main/kotlin/com/apprain/barcode_scanner/ScannerPlatformView.kt
package com.apprain.barcode_scanner

import android.app.Activity
import android.content.Context
import android.util.Log
import android.view.View
import android.widget.FrameLayout
import androidx.camera.core.*
import androidx.camera.lifecycle.ProcessCameraProvider
import androidx.camera.view.PreviewView
import androidx.core.content.ContextCompat
import androidx.lifecycle.LifecycleOwner
import com.google.mlkit.vision.barcode.BarcodeScanner
import com.google.mlkit.vision.barcode.BarcodeScannerOptions
import com.google.mlkit.vision.barcode.BarcodeScanning
import com.google.mlkit.vision.barcode.common.Barcode
import com.google.mlkit.vision.common.InputImage
import io.flutter.plugin.common.EventChannel
import io.flutter.plugin.platform.PlatformView
import java.util.concurrent.Executors
import java.util.concurrent.atomic.AtomicBoolean

/**
 * Native Android view that hosts CameraX preview + ML Kit barcode analysis.
 *
 * Key design decisions:
 * - Uses [PreviewView] for zero-copy camera preview rendering.
 * - [ImageAnalysis] with STRATEGY_KEEP_ONLY_LATEST for back-pressure handling.
 * - [AtomicBoolean] gate ensures only one frame is processed at a time.
 * - Supports ROI pre-crop for focused scanning.
 * - Frame throttling via configurable maxFrameRate.
 */
class ScannerPlatformView(
    private val context: Context,
    private val activity: Activity?,
    private val viewId: Int,
    private val creationParams: Map<*, *>?,
    private val scanEventChannel: EventChannel,
    private val cameraEventChannel: EventChannel,
) : PlatformView {

    companion object {
        private const val TAG = "ApprainScanner"
    }

    // ── View ─────────────────────────────────────────────────────────────────

    private val container = FrameLayout(context)
    private val previewView = PreviewView(context).apply {
        implementationMode = PreviewView.ImplementationMode.PERFORMANCE
        scaleType = PreviewView.ScaleType.FILL_CENTER
    }

    // ── Camera ───────────────────────────────────────────────────────────────

    private var cameraProvider: ProcessCameraProvider? = null
    private var camera: Camera? = null
    private var preview: Preview? = null
    private var imageAnalysis: ImageAnalysis? = null
    private val analysisExecutor = Executors.newSingleThreadExecutor()
    private val isProcessing = AtomicBoolean(false)
    private var lastAnalysisTimestamp = 0L

    // ── ML Kit ───────────────────────────────────────────────────────────────

    private var barcodeScanner: BarcodeScanner? = null

    // ── Config ───────────────────────────────────────────────────────────────

    private var useFrontCamera = false
    private var enableScanWindow = true
    private var scanWindowWidthRatio = 0.75
    private var scanWindowHeightRatio = 0.35
    private var scanWindowVerticalOffsetRatio = 0.10
    private var enablePreprocessing = true
    private var maxFrameRate = 15
    private var enableTorch = false

    // ── Event Sinks ──────────────────────────────────────────────────────────

    private var scanEventSink: EventChannel.EventSink? = null
    private var cameraEventSink: EventChannel.EventSink? = null

    init {
        container.addView(previewView)
        setupEventChannels()
    }

    override fun getView(): View = container

    override fun dispose() {
        disposeResources()
    }

    // ── Public API (called from Plugin via MethodChannel) ────────────────────

    fun initialize(config: Map<*, *>?) {
        applyConfig(config)
        initBarcodeScanner(config)
    }

    fun startCamera() {
        val cameraProviderFuture = ProcessCameraProvider.getInstance(context)
        cameraProviderFuture.addListener({
            try {
                cameraProvider = cameraProviderFuture.get()
                bindCameraUseCases()
                cameraEventSink?.success(mapOf("type" to "ready"))
                Log.d(TAG, "Camera started successfully.")
            } catch (e: Exception) {
                Log.e(TAG, "Camera start failed: ${e.message}", e)
                cameraEventSink?.success(mapOf("type" to "error", "message" to (e.message ?: "Unknown error")))
            }
        }, ContextCompat.getMainExecutor(context))
    }

    fun stopCamera() {
        cameraProvider?.unbindAll()
        camera = null
        Log.d(TAG, "Camera stopped.")
    }

    fun disposeResources() {
        stopCamera()
        barcodeScanner?.close()
        barcodeScanner = null
        analysisExecutor.shutdown()
        scanEventSink = null
        cameraEventSink = null
        Log.d(TAG, "Resources disposed.")
    }

    fun setTorch(enabled: Boolean) {
        camera?.cameraControl?.enableTorch(enabled)
        enableTorch = enabled
    }

    fun setZoom(zoomLevel: Float) {
        camera?.cameraControl?.setLinearZoom(zoomLevel.coerceIn(0f, 1f))
    }

    fun setFocusPoint(x: Float, y: Float) {
        val factory = previewView.meteringPointFactory
        val point = factory.createPoint(
            x * previewView.width,
            y * previewView.height,
        )
        val action = FocusMeteringAction.Builder(point)
            .setAutoCancelDuration(3, java.util.concurrent.TimeUnit.SECONDS)
            .build()
        camera?.cameraControl?.startFocusAndMetering(action)
    }

    fun switchCamera() {
        useFrontCamera = !useFrontCamera
        if (cameraProvider != null) {
            bindCameraUseCases()
        }
    }

    fun updateConfig(config: Map<*, *>?) {
        applyConfig(config)
    }

    // ── Private Implementation ───────────────────────────────────────────────

    private fun setupEventChannels() {
        scanEventChannel.setStreamHandler(object : EventChannel.StreamHandler {
            override fun onListen(arguments: Any?, events: EventChannel.EventSink?) {
                scanEventSink = events
            }
            override fun onCancel(arguments: Any?) {
                scanEventSink = null
            }
        })

        cameraEventChannel.setStreamHandler(object : EventChannel.StreamHandler {
            override fun onListen(arguments: Any?, events: EventChannel.EventSink?) {
                cameraEventSink = events
            }
            override fun onCancel(arguments: Any?) {
                cameraEventSink = null
            }
        })
    }

    private fun applyConfig(config: Map<*, *>?) {
        config ?: return
        useFrontCamera = config["useFrontCamera"] as? Boolean ?: useFrontCamera
        enableScanWindow = config["enableScanWindow"] as? Boolean ?: enableScanWindow
        scanWindowWidthRatio = (config["scanWindowWidthRatio"] as? Number)?.toDouble() ?: scanWindowWidthRatio
        scanWindowHeightRatio = (config["scanWindowHeightRatio"] as? Number)?.toDouble() ?: scanWindowHeightRatio
        scanWindowVerticalOffsetRatio = (config["scanWindowVerticalOffsetRatio"] as? Number)?.toDouble() ?: scanWindowVerticalOffsetRatio
        enablePreprocessing = config["enablePreprocessing"] as? Boolean ?: enablePreprocessing
        maxFrameRate = (config["maxFrameRate"] as? Number)?.toInt() ?: maxFrameRate
        enableTorch = config["enableTorch"] as? Boolean ?: enableTorch
    }

    private fun initBarcodeScanner(config: Map<*, *>?) {
        val allowedFormats = (config?.get("allowedFormats") as? List<*>)
            ?.mapNotNull { it as? String }
            ?: emptyList()

        val options = if (allowedFormats.isNotEmpty()) {
            val formats = allowedFormats.mapNotNull { formatStringToInt(it) }
            if (formats.size >= 2) {
                BarcodeScannerOptions.Builder()
                    .setBarcodeFormats(formats[0], *formats.drop(1).toIntArray())
                    .build()
            } else if (formats.size == 1) {
                BarcodeScannerOptions.Builder()
                    .setBarcodeFormats(formats[0])
                    .build()
            } else {
                BarcodeScannerOptions.Builder().build()
            }
        } else {
            // Scan all formats by default
            BarcodeScannerOptions.Builder().build()
        }

        barcodeScanner?.close()
        barcodeScanner = BarcodeScanning.getClient(options)
    }

    private fun bindCameraUseCases() {
        val provider = cameraProvider ?: return
        provider.unbindAll()

        val lifecycleOwner = activity as? LifecycleOwner ?: return

        val cameraSelector = if (useFrontCamera) {
            CameraSelector.DEFAULT_FRONT_CAMERA
        } else {
            CameraSelector.DEFAULT_BACK_CAMERA
        }

        // Preview use case
        preview = Preview.Builder()
            .build()
            .also { it.surfaceProvider = previewView.surfaceProvider }

        // Image Analysis use case
        imageAnalysis = ImageAnalysis.Builder()
            .setBackpressureStrategy(ImageAnalysis.STRATEGY_KEEP_ONLY_LATEST)
            .setTargetResolution(android.util.Size(1280, 720))
            .build()
            .also { analysis ->
                analysis.setAnalyzer(analysisExecutor) { imageProxy ->
                    processFrame(imageProxy)
                }
            }

        try {
            camera = provider.bindToLifecycle(
                lifecycleOwner,
                cameraSelector,
                preview,
                imageAnalysis,
            )

            // Apply torch state
            if (enableTorch) {
                camera?.cameraControl?.enableTorch(true)
            }

            Log.d(TAG, "Camera use cases bound. Front: $useFrontCamera")
        } catch (e: Exception) {
            Log.e(TAG, "Failed to bind camera use cases: ${e.message}", e)
            cameraEventSink?.success(mapOf("type" to "error", "message" to (e.message ?: "Bind failed")))
        }
    }

    @androidx.camera.core.ExperimentalGetImage
    private fun processFrame(imageProxy: ImageProxy) {
        // Frame throttling
        val now = System.currentTimeMillis()
        val minInterval = if (maxFrameRate > 0) 1000L / maxFrameRate else 0L
        if (now - lastAnalysisTimestamp < minInterval) {
            imageProxy.close()
            return
        }

        // Single-frame gate — prevent concurrent processing
        if (!isProcessing.compareAndSet(false, true)) {
            imageProxy.close()
            return
        }

        lastAnalysisTimestamp = now

        val mediaImage = imageProxy.image
        if (mediaImage == null) {
            isProcessing.set(false)
            imageProxy.close()
            return
        }

        val inputImage = InputImage.fromMediaImage(
            mediaImage,
            imageProxy.imageInfo.rotationDegrees,
        )

        val scanner = barcodeScanner
        if (scanner == null) {
            isProcessing.set(false)
            imageProxy.close()
            return
        }

        scanner.process(inputImage)
            .addOnSuccessListener { barcodes ->
                if (barcodes.isNotEmpty()) {
                    val results = barcodes
                        .filter { barcode ->
                            if (!enableScanWindow) return@filter true
                            // Post-processing ROI filter: check if barcode center
                            // falls within the scan window
                            val box = barcode.boundingBox ?: return@filter true
                            val imageWidth = inputImage.width.toDouble()
                            val imageHeight = inputImage.height.toDouble()
                            val centerX = box.centerX() / imageWidth
                            val centerY = box.centerY() / imageHeight

                            val windowLeft = (1.0 - scanWindowWidthRatio) / 2.0
                            val windowRight = 1.0 - windowLeft
                            val windowCenterY = 0.5 - scanWindowVerticalOffsetRatio
                            val windowTop = windowCenterY - scanWindowHeightRatio / 2.0
                            val windowBottom = windowCenterY + scanWindowHeightRatio / 2.0

                            centerX in windowLeft..windowRight && centerY in windowTop..windowBottom
                        }
                        .mapNotNull { barcode ->
                            val rawValue = barcode.rawValue ?: return@mapNotNull null
                            mapOf(
                                "rawValue" to rawValue,
                                "format" to formatIntToString(barcode.format),
                                "corners" to (barcode.cornerPoints?.flatMap {
                                    listOf(it.x.toDouble(), it.y.toDouble())
                                } ?: emptyList<Double>()),
                                "confidence" to 1.0,
                            )
                        }

                    if (results.isNotEmpty()) {
                        ContextCompat.getMainExecutor(context).execute {
                            scanEventSink?.success(results)
                        }
                    }
                }
            }
            .addOnCompleteListener {
                isProcessing.set(false)
                imageProxy.close()
            }
    }

    // ── Format Mapping ───────────────────────────────────────────────────────

    private fun formatStringToInt(format: String): Int? = when (format.uppercase()) {
        "QR_CODE" -> Barcode.FORMAT_QR_CODE
        "CODE_128" -> Barcode.FORMAT_CODE_128
        "CODE_39" -> Barcode.FORMAT_CODE_39
        "CODE_93" -> Barcode.FORMAT_CODE_93
        "EAN_13" -> Barcode.FORMAT_EAN_13
        "EAN_8" -> Barcode.FORMAT_EAN_8
        "UPC_A" -> Barcode.FORMAT_UPC_A
        "UPC_E" -> Barcode.FORMAT_UPC_E
        "DATA_MATRIX" -> Barcode.FORMAT_DATA_MATRIX
        "PDF417" -> Barcode.FORMAT_PDF417
        "AZTEC" -> Barcode.FORMAT_AZTEC
        "CODABAR" -> Barcode.FORMAT_CODABAR
        "ITF" -> Barcode.FORMAT_ITF
        else -> null
    }

    private fun formatIntToString(format: Int): String = when (format) {
        Barcode.FORMAT_QR_CODE -> "QR_CODE"
        Barcode.FORMAT_CODE_128 -> "CODE_128"
        Barcode.FORMAT_CODE_39 -> "CODE_39"
        Barcode.FORMAT_CODE_93 -> "CODE_93"
        Barcode.FORMAT_EAN_13 -> "EAN_13"
        Barcode.FORMAT_EAN_8 -> "EAN_8"
        Barcode.FORMAT_UPC_A -> "UPC_A"
        Barcode.FORMAT_UPC_E -> "UPC_E"
        Barcode.FORMAT_DATA_MATRIX -> "DATA_MATRIX"
        Barcode.FORMAT_PDF417 -> "PDF417"
        Barcode.FORMAT_AZTEC -> "AZTEC"
        Barcode.FORMAT_CODABAR -> "CODABAR"
        Barcode.FORMAT_ITF -> "ITF"
        else -> "UNKNOWN"
    }
}
